### PR TITLE
New Resource: `azurerm_function_app`

### PR DESCRIPTION
**Pull request by JunyiYi**
* Add `primary_connection_string` and `secondary_connection_string` to `azurerm_storage_account` resource
* Add `azurerm_function_app` resource to the provider
* Add/Update documentations and tests

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
Fixed #610

File: azurerm/resource_arm_function_app_test.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMFunctionApp_basic(t *testing.T) {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_function_app_test.go]: can we add another acceptance test for going from `V1 -> Beta` (and back)? It's possible to use multiple test steps to do this and assert on the value of the `version` field in each step - as we do here: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_traffic_manager_endpoint_test.go#L111

File: azurerm/resource_arm_function_app.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["sites"]
+
+	if d.HasChange("app_settings") {
+		appSettings := expandAppServiceAppSettings(d)
+		settings := web.StringDictionary{
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_function_app.go]: Exactly! Thanks for finding the bug.

File: azurerm/resource_arm_function_app.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	d.Set("location", azureRMNormalizeLocation(*resp.Location))
+
+	if props := resp.SiteProperties; props != nil {
+		d.Set("app_service_plan_id", props.ServerFarmID)
+		d.Set("enabled", props.Enabled)
+		d.Set("default_hostname", props.DefaultHostName)
+	}
+
+	appSettings := flattenAppServiceAppSettings(appSettingsResp.Properties)
+	if err := d.Set("app_settings", appSettings); err != nil {
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_function_app.go]: Absolutely.

File: azurerm/resource_arm_storage_account.go [hunk 4]
```diff
@@ -526,6 +536,16 @@ func resourceArmStorageAccountRead(d *schema.ResourceData, meta interface{}) err
 		d.Set("primary_location", props.PrimaryLocation)
 		d.Set("secondary_location", props.SecondaryLocation)
 
+		if len(accessKeys) > 0 {
+			pcs := fmt.Sprintf("DefaultEndpointsProtocol=https;AccountName=%s;AccountKey=%s;EndpointSuffix=core.windows.net", *resp.Name, *accessKeys[0].Value)
+			d.Set("primary_connection_string", pcs)
+		}
+
+		if len(accessKeys) > 1 {
+			scs := fmt.Sprintf("DefaultEndpointsProtocol=https;AccountName=%s;AccountKey=%s;EndpointSuffix=core.windows.net", *resp.Name, *accessKeys[1].Value)
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_storage_account.go]: Nice suggestion.

File: azurerm/resource_arm_function_app_test.go (hunk 1 above)
Comment (JunyiYi) [reply azurerm/resource_arm_function_app_test.go]: Sure thing.

File: website/docs/r/function_app.html.markdown [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+description: |-
+  Manages an Azure Functions service.
+
+---
+
+# azurerm_function_app
+
+Manages an Azure Functions service.
+
+## Example Usage
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply website/docs/r/function_app.html.markdown]: Maybe `Example Usage (with App Service Plan)` is more accurate.

File: website/docs/r/function_app.html.markdown [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+* `tags` - (Optional) A mapping of tags to assign to the resource. Changing this forces a new resource to be created.
+
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The ID of the Function App
+
+* `default_hostname` - The default hostname associated with the Function App - such as `mysite.azurewebsites.net`
```
Comment (tombuildsstuff) [inline website/docs/r/function_app.html.markdown]: going to push an update to add Import documentation here

File: website/docs/r/function_app.html.markdown [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+* `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
+
+* `app_service_plan_id` - (Required) The ID of the App Service Plan within which to create this Function App. Changing this forces a new resource to be created.
+
+* `storage_connection_string` - (Required) The connection string of the backend storage account which will be used by this Function App (such as the dashboard, logs).
+
+* `app_settings` - (Optional) A key-value pair of App Settings.
+
+* `enabled` - (Optional) Is the Azure Function service enabled? Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/function_app.html.markdown]: going to push an update to change this to `Function App`

File: website/docs/r/function_app.html.markdown (hunk 6 above)
Comment (tombuildsstuff) [reply website/docs/r/function_app.html.markdown]: (done)

File: website/docs/r/function_app.html.markdown (hunk 7 above)
Comment (tombuildsstuff) [reply website/docs/r/function_app.html.markdown]: (done)

**Comment by tombuildsstuff** [comment]
(after testing this - I've also made the field `storage_connection_string` sensitive, since it's a connection string)

**Comment by tombuildsstuff** [comment]
@JunyiYi thanks for all the work on this - this LGTM 👍

**Comment by tombuildsstuff** [comment]
Whilst verifying this I've ended up pushing a couple of tests to add/update App Settings & Tags - which pass:

screen shot 2018-01-11 at 16 00 21

**Pull request merged**